lib/galaxy/model/__init__.py

```diff
@@ -5840,7 +5844,8 @@ def finalize(self, collection_type_description):
     def dataset_instances(self):
         db_session = object_session(self)
         if db_session and self.id:
-            return self._get_nested_collection_attributes(return_entities=(HistoryDatasetAssociation,)).all()
+            statement = self._get_nested_collection_attributes(return_entities=(HistoryDatasetAssociation,))
+            return db_session.execute(statement).all()
```

Comment:
```suggestion
            return db_session.execute(statement).scalars().all()
```
So, this will fix one error (if you look at the `test_model_discovery` code, `dataset0/1` (lines 203-204) are not HDAs, but tuples containing HDAs. The `scalars()` call fixes that (it's new in 1.4). That's just one thing - other stuff is broken. I'm thinking, maybe the move from `query` to `select` (which seems nontrivial given the abundance of such low-level code in our code base) might push us to come up with better abstractions for such code in this module (at least).